File: backend/src/main/java/com/matimeline/eventmanager/domain/models/Category.java
package com.matimeline.eventmanager.domain.models;

import java.util.UUID;

public class Category {
    private UUID id;
    private String name;
    private String color;
    private String description;
    // #52 (ADR-002) : propriétaire de la catégorie. NULL == catégorie « système »
    // (lisible de tous, non modifiable — cf. V8__category_ownership.sql).
    private UUID ownerId;

    public Category(UUID id, String name) {
        this.id = id;
        this.name = name;
    }

    public Category(UUID id, String name, String color, String description) {
        this.id = id;
        this.name = name;
        this.color = color;
        this.description = description;
    }

    public Category(UUID id, String name, String color, String description, UUID ownerId) {
        this(id, name, color, description);
        this.ownerId = ownerId;
    }

    public UUID getId() {
        return id;
    }

    public String getName() {
        return name;
    }

    public void setName(String name) {
        this.name = name;
    }

    public String getColor() {
        return color;
    }

    public void setColor(String color) {
        this.color = color;
    }

    public String getDescription() {
        return description;
    }

    public void setDescription(String description) {
        this.description = description;
    }

    public UUID getOwnerId() {
        return ownerId;
    }

    public void setOwnerId(UUID ownerId) {
        this.ownerId = ownerId;
    }
}

File: frontend/src/types/category.ts
import { z } from 'zod'

/**
 * #65 / S10 #52 — DTO `CategoryResponse` du backend.
 *
 * Sync Zod ↔ DTO (cf. cp-frontend « Sync Zod ↔ DTO ») :
 *   - `id`   : UUID (string).
 *   - `name` : libellé affiché.
 *   - `system` : booléen. `true` = catégorie système (owner NULL côté backend,
 *     ADR-002) → lisible de tous mais NON supprimable (403). Le DTO n'expose
 *     JAMAIS l'ownerId, uniquement ce booléen dérivé (br-categories).
 *   - `color` (#61) : couleur de la catégorie (`CategoryResponse.color`). Utilisée
 *     par le `ProductDrawer` comme couleur héritée par défaut du produit. Backend
 *     `Category.getColor()` peut être `null` → `.nullable().optional()` (défensif,
 *     et rétro-compatible avec les fixtures #65 qui l'omettent).
 *
 * `id`/`name`/`system` : champs backend toujours présents → pas d'`.optional()`.
 */
export const categorySchema = z.object({
  id: z.string(),
  name: z.string(),
  system: z.boolean(),
  color: z.string().nullable().optional(),
  description: z.string().nullable().optional(),
})

export type Category = z.infer<typeof categorySchema>

/**
 * #62 — Payloads d'écriture catégorie (POST/PATCH `/api/categories`).
 *
 * Sync Zod ↔ DTO backend (`CategoryRequest` / `CategoryUpdateRequest`) :
 *   - `name`        : `@NotBlank @Size(max=255)` (BR-CAT-001) → `min(1).max(255)`.
 *   - `color`       : `@Size(max=255)` optionnel. Le backend N'IMPOSE PAS le format
 *     hex (contrairement aux produits `@Pattern`) : le picker libre (react-colorful)
 *     émet `#RRGGBB`, mais on ne sur-contraint PAS le contrat → simple string bornée.
 *   - `description` : `@Size(max=255)` optionnelle.
 *
 * Le DTO n'a PAS de sémantique `clearColor` (contrairement aux produits #158) :
 * PATCH écrase `color`/`description` avec la valeur portée (y compris `null`/vide
 * pour effacer). On envoie donc toujours name (+ color/description tels quels).
 */
export const categoryCreateSchema = z.object({
  name: z.string().min(1).max(255),
  color: z.string().max(255).optional(),
  description: z.string().max(255).optional(),
})

export type CategoryCreate = z.infer<typeof categoryCreateSchema>

/**
 * PATCH : le backend porte `name` obligatoire (BR-CAT-001/003, `@NotBlank`).
 * `color`/`description` optionnels côté transport (absent = inchangé côté drawer,
 * qui n'envoie que le diff).
 */
export const categoryUpdateSchema = z.object({
  name: z.string().min(1).max(255),
  color: z.string().max(255).optional(),
  description: z.string().max(255).optional(),
})

export type CategoryUpdate = z.infer<typeof categoryUpdateSchema>

/**
 * Factory i18n du schéma formulaire (messages traduits, cf. cp-frontend « deux
 * familles de schémas »). Seul `name` porte une validation bloquante côté UI
 * (BR-CAT-001) ; color/description sont libres. Utilisé par le `zodResolver` du
 * `CategoryDrawer`.
 */
export function createCategoryFormSchema(t: (key: string) => string) {
  return z.object({
    name: z.string().min(1, { message: t('nameRequired') }).max(255, { message: t('nameTooLong') }),
    color: z.string().max(255).optional(),
    description: z.string().max(255).optional(),
  })
}

export type CategoryFormValues = z.infer<ReturnType<typeof createCategoryFormSchema>>
